proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,175 @@
+---
+simd: '0024'
+title: Automatially repair and start for a cluster restart
+authors:
+  - Wen Xu (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-04-07
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Improve the current [cluster restart]
+(https://docs.solana.com/running-validator/restart-cluster)
+procedure such that validators can automatically figure out confirmed slots
+and then proceed to restart if everything looks fine.
+
+## Motivation
+
+Currently during a cluster restart, validator operators need to decide latest
+optimistically confirmed slot, then restart the validators with new commandline
+arguments.
+
+The current process involves a lot of human intervention, if people make a
+mistake in deciding the highest optimistically confirmed slot, it could mean
+rollback of user transactions after those transactions have been confirmed,
+which is not acceptable.
+
+We aim to automate the finding of highest optimistically confirmed slot and
+block data distribution, so that we can lower the possibility of human mistakes
+in the cluster restart process.
+
+## Alternatives Considered
+
+See [Handling Common Solana Outages]
+(https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc)
+for details.
+
+There are many proposals about automatically detecting that the cluster is
+in an outage so validators should enter a recovery process automatically.
+
+While getting human out of the loop greatly improves recovery speed,
+automaticlly restarting the whole cluster still seems risky. Because if
+the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+## New Terminology
+
+* SilentRepairMode - when validators restart in this new mode, they will
+talk with each other to find the highest optimistically confirmed slot and
+repair all necessary blocks. To improve speed and guarantee simplicity,
+Turbine, vote, and new block generation are paused in this mode.
+
+## Detailed Design
+
+The new protocol tries to make all 80% restarting validators get the same
+data blocks and the same set of last votes among them, then they can probably
+make the same decision and then proceed.
+
+The steps roughly look like this:
+
+1. Everyone freezes, no new blocks, no new votes, and no Turbine
+
+2. Make all blocks which can potentially have been optimistically confirmed
+before the freeze propagate to everyone
+
+3. Make restart participants' last votes before the freeze propagate to
+everyone
+
+4. Now see if enough people can optimistically agree on one block (same slot
+and hash) to restart from
+
+4.1 If yes, proceed and restart
+
+4.2 If no, freeze and print out what you think is wrong, wait for human
+
+A new command line arg --RepairAndRestart <optional_slots_to_send> is added.
```

Comment:
I'm still hoping we can avoid this `optional_slots_to_send` argument. What if we always sent N hours (where N is in the single digits) of slots before the highest root the validator has observed, plus all slots above its root.

Reply:
It's fine to remove it, just here for safety. We can set the default to 9 hours, which should be enough time?

Reply:
Assuming the `LastVotedForkSlots` value is fixed size, we could just full it with all recent slots the validator has voted on. That'd include some amount of slots over its highest root, and then however many slots below that root that can fit  rather than prescribing an exact time bound (but yeah 9-12 hours will be more than enough based on all past outages and restarts). 

But so if the validator has voted on too many slots over the highest root such that the data size of `LastVotedForkSlots` would be exceeded, that validator can't participate in this style of automated restart.  If too many validators are in this condition then the operators need to fall back to the current method of identifying the restart slot via social consensus.

Reply:
Sounds good, I put this in the next section.